[FLINK-21445] Refactors the PackagedProgramRetriever implementation and adds configuration to PackagedProgram

XComp opened the pull request:
## What is the purpose of the change

The initial purpose was to add the configuration to the `PackagedProgram`. As part of this effort, we thought of refactoring the code a bit to improve testability. See the discussions in #15020.

## Brief change log

* The extraction of job class information is moved into its own interface `EntryClassInformationProvider`
* Two implementations are provided:
  * `FromJarEntryClassInformationProvider` is a basic implementation expecting some jar to be present. This replaces the code where a jar is provided or where the job is actually PyFlink-based
  * `FromClasspathEntryClassInformationProvider` access either the passed user classpath or the system classpath to determine the entry class (this replaces the fallback logic which was provided in the old implementation's `ClassPathPackagedProgramRetriever` and the singleton `ClassPathPackagedProgramRetriever.JarsOnClassPath` which was used for loading the information for `java.class.path`.
* `DefaultPackagedProgramRetriever` is introduced as a generic implementation of the already available `PackagedProgramRetriever` interface.
* The fix for [FLINK-21445](https://issues.apache.org/jira/browse/FLINK-21445) was added in a separate commit

## Verifying this change

* The old tests were moved from `ClassPathPackagedProgramRetrieverTest` into `DefaultPackagedProgramRetrieverTest` to make sure that the changes cover the already available functionality
* `ClasspathProvider` was introduced implementing `ExternalResource` to encapsulate the logic which was initially provided in `ClassPathPackagedProgramRetrieverTest.init`
* Additionally, tests were added to check `DefaultPackagedProgramRetriever.createEntryClassInformationProvider`
* Explicit testing was added for the `EntryClassInformationProvider` implementations through `FromJarEntryClassInformationProviderTest` and `FromClasspathEntryClassInformationProviderTest`
* `DefaultPackagedProgramRetrieverTest.testConfigurationIsConsidered` was added to cover the fix for [FLINK-21445](https://issues.apache.org/jira/browse/FLINK-21445)
* `DefaultPackagedProgramRetrieverTest.testChildFirstDefaultConfiguration` was added as a sanity check to show that the default configuration would lead to a different classloader being used

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

SteNicholas commented on flink-clients/src/main/java/org/apache/flink/client/program/PackagedProgramRetrieverImpl.java:
```diff
@@ -0,0 +1,203 @@
+            @Nullable File userLibDir,
+            @Nullable File jarFile,
+            @Nullable String jobClassName,
+            String[] programArgs)
```
Does this need to add `Configuration` parameter?

XComp replied:
Thanks for being so responsive looking into the PR. You're right. The draft is still a draft. I'm gonna add the fixes for configuration and savepoint settings soon. I needed to create the draft first because I wanted to link it to [FLINK-23154](https://issues.apache.org/jira/browse/FLINK-23154).

SteNicholas commented on flink-yarn/src/main/java/org/apache/flink/yarn/entrypoint/YarnApplicationClusterEntryPoint.java:
```diff
@@ -129,16 +129,12 @@ private static PackagedProgramRetriever getPackagedProgramRetriever(
-                        .setJarFile(userApplicationJar)
-                        .setJobClassName(jobClassName);
-        return retrieverBuilder.build();
+        return PackagedProgramRetrieverImpl.create(
```
This doesn't pass the `Configuration`.

SteNicholas commented on flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneApplicationClusterEntryPoint.java:
```diff
@@ -111,13 +111,9 @@ private static PackagedProgram getPackagedProgram(
-                        .setUserLibDirectory(userLibDir)
-                        .setJobClassName(jobClassName);
-        return retrieverBuilder.build();
+        return PackagedProgramRetrieverImpl.create(userLibDir, jobClassName, programArguments);
```
This doesn't pass the `Configuration`.

SteNicholas commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/entrypoint/KubernetesApplicationClusterEntrypoint.java:
```diff
@@ -105,22 +105,20 @@ private static PackagedProgramRetriever getPackagedProgramRetriever(
                     KubernetesUtils.checkJarFileForApplicationMode(configuration);
             Preconditions.checkArgument(pipelineJars.size() == 1, "Should only have one jar");
-            retrieverBuilder.setJarFile(pipelineJars.get(0));
+            return PackagedProgramRetrieverImpl.create(
```
This doesn't pass the `Configuration`.

SteNicholas commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/entrypoint/KubernetesApplicationClusterEntrypoint.java:
```diff
@@ -105,22 +105,20 @@ private static PackagedProgramRetriever getPackagedProgramRetriever(
         }
-        return retrieverBuilder.build();
+
+        return PackagedProgramRetrieverImpl.create(userLibDir, jobClassName, programArguments);
```
This doesn't pass the `Configuration`.

XComp commented on flink-clients/src/test/java/org/apache/flink/client/program/PackagedProgramRetrieverImplTest.java:
```diff
@@ -0,0 +1,590 @@
+    }
+
+    @Test
+    // TODO: we might want to change this behavior triggering a failure here
```
Is this a right concern?

zentol replied:
I don't understand the comment, please elaborate.

XComp replied:
Right now, we can specify a class and don't check whether it's actual present on the classpath. I was wondering whether that's something we want to change. We're already checking whether the class is available in the user classpath. We could do the same for the system classpath.

XComp commented on flink-clients/src/test/java/org/apache/flink/client/program/PackagedProgramRetrieverImplTest.java:
```diff
@@ -0,0 +1,590 @@
+    @Test
+    public void testJobGraphRetrievalJobClassNameHasPrecedenceOverClasspath()
+            throws IOException, FlinkException, ProgramInvocationException {
+        // TODO: I don't get this test: Shouldn't the related jar be on the classpath?
+        //        classpathProvider.setSystemClasspathWithTestJob();
```
This test was migrated from the old `ClassPathPackagedProgramRetrieverTest`. But it feels to me that we don't want that behavior, do we?

XComp commented on flink-clients/src/test/java/org/apache/flink/client/testjar/ClasspathProvider.java:
```diff
@@ -0,0 +1,221 @@
+import java.util.stream.StreamSupport;
+
+/** {@code ClasspathProvider} offers methods to generate classpaths based on actual jars. */
+public class ClasspathProvider extends ExternalResource {
```
I moved the logic which was included in the legacy `ClassPathPackagedProgramRetrieverTest` into its own class making it available as a `@Rule` annotated member

XComp replied:
I refactored the class to remove the "hodgepodge". Now, one instance serves as a utility method provider for a specific case.

XComp commented on flink-clients/src/test/java/org/apache/flink/client/program/PackagedProgramRetrieverImplTest.java:
```diff
@@ -0,0 +1,557 @@
+    }
+
+    @Test
+    public void testJobGraphRetrievalJobClassNameHasPrecedenceOverClasspath()
```
I left the test here for now for transparency reasons. I think we should remove it but make the job fail in case the job class is not even provided on the system class path.

wangyang0918 replied:
Yes. The legacy `ClassPathPackagedProgramRetriever.Builder#setJarsOnClassPath` is never used in production code. And it seems not have some benefits to specify the jar name on classpath instead of entry class.

+1 to remove this test.

zentol replied:
It doesn't matter where the setter is used in production or not, the purpose of the test is to ensure behavior despite _changes_ in production code (i.e., if the setter were used).

Where else do you test that, if 2 entrypoints are on the CP and the user specifies the entrypoint class, that the user argument is honored?

XComp replied:
This test only succeeds in the old implementation because we're not checking whether the given job class name is on the system classpath. But @zentol is right with checking that the parameters are acknowledged for this specific case. Although, it should be moved to `PackagedProgramTest`, I guess. But I leave it here (and just rename it) because I feel that that refactoring the testcases in this matter is out of scope of this PR.

XComp commented:
@SteNicholas @wangyang0918 I updated the PR. It would be nice to get your thoughts on that one.

SteNicholas commented on flink-clients/src/main/java/org/apache/flink/client/deployment/application/EntryClassInformationProvider.java:
```diff
@@ -0,0 +1,45 @@
+
+/**
+ * {@code EntryClassInformationProvider} provides information about the entry class that shall be
+ * used by the {@link org.apache.flink.client.program.PackagedProgramRetrieverImpl}.
```
Could `{@link org.apache.flink.client.program.PackagedProgramRetrieverImpl}` update to `{@link PackagedProgramRetrieverImpl}`? In my pull request, this is recommended to update like this.

SteNicholas commented on flink-clients/src/main/java/org/apache/flink/client/deployment/application/FromClasspathEntryClassInformationProvider.java:
```diff
@@ -0,0 +1,207 @@
+                ExceptionUtils.rethrow(
+                        e,
+                        String.format(
+                                "Failed to open user class path %s. Make sure that all files on the user classpath can be accessed.",
```
```suggestion
                                "Failed to open user classpath %s. Make sure that all files on the user classpath can be accessed.",
```

SteNicholas commented on flink-clients/src/main/java/org/apache/flink/client/deployment/application/FromJarEntryClassInformationProvider.java:
```diff
@@ -0,0 +1,76 @@
+public class FromJarEntryClassInformationProvider implements EntryClassInformationProvider {
+
+    private final File jarFile;
+    private final String jobClassName;
```
Does this need to inherit `FromClasspathEntryClassInformationProvider`?

XComp replied:
Where do you see the benefits of using inheritance here?

SteNicholas commented on flink-clients/src/main/java/org/apache/flink/client/deployment/application/JarManifestParser.java:
```diff
@@ -118,7 +118,7 @@ static JarFileWithEntryClass findOnlyEntryClass(Iterable<File> jarFiles) throws
      * @throws IOException If there is an error accessing the JAR
      */
-    private static Optional<String> findFirstManifestAttribute(File jarFile, String... attributes)
+    public static Optional<String> findFirstManifestAttribute(File jarFile, String... attributes)
```
Why not `findFirstManifestAttribute` define as `protected`?

SteNicholas commented:
@XComp , thanks for your efforts to update the pull request again me. I have reviewed the changes of the refactors. I have confusion about why not validate whether to retrever job graph in tests, which originally validate through `retrieveJobGraph`.

XComp commented:
> @XComp , thanks for your efforts to update the pull request again me. I have reviewed the changes of the refactors. I have confusion about why not validate whether to retrever job graph in tests, which originally validate through retrieveJobGraph.

Thanks for your review. I addressed two of the changes and rebased the branch. About your questions with not using the retriever job graph. Building the `PackagedProgram` will trigger the class loading internally. This is problematic for testing the system classpath functionality.
I tried to use legacy test code as much as possible. New tests, in contrast, avoid loading the class from classpath but rather check whether the correct class is selected. That's why the test implementation focuses on the `EntryClassInformationProvider`. It's more of a unit test rather than a test testing multiple classes interacting with each other.

SteNicholas commented:
LGTM.

wangyang0918 commented on flink-clients/src/test/java/org/apache/flink/client/program/PackagedProgramRetrieverImplTest.java:
```diff
@@ -285,6 +287,26 @@ public void testSavepointRestoreSettings()
+            fail("This case should throw exception !");
+        } catch (FlinkException e) {
+            assertThat(
+                    ExceptionUtils.findThrowableWithMessage(
```
I think we could use `assertThat(e, FlinkMatchers.containsMessage(msg))` here.

wangyang0918 commented on flink-clients/src/test/java/org/apache/flink/client/program/PackagedProgramRetrieverImplTest.java:
```diff
@@ -372,6 +372,35 @@ public void testRetrieveCorrectUserClasspathsWithoutSpecifiedEntryClass()
+                jobGraph.getClasspaths().stream().map(URL::toString).collect(Collectors.toList());
+
+        final Path workingDirectory = FileUtils.getCurrentWorkingDirectory();
+        final List<String> expectedClasspath = new ArrayList<>();
+        for (File file : classpathProvider.getUserDirectoryWithOneEntryClass().listFiles()) {
+            if (!file.getName().endsWith("jar")) {
+                // only jars are expected
+                continue;
+            }
+
+            Path relativePath = FileUtils.relativizePath(workingDirectory, file.toPath());
+            expectedClasspath.add(FileUtils.toURL(relativePath).toString());
+        }
```
Could be deduplicated by introducing a new method.

wangyang0918 commented on flink-clients/src/test/java/org/apache/flink/client/deployment/application/FromJarEntryClassInformationProviderTest.java:
```diff
@@ -72,9 +73,13 @@ public void testEitherJobClassNameOrJarHasToBeSet() {
+        // TODO: see FLINK-23154
+        Assume.assumeTrue(
+                "This tests only succeeds if " + ConfigConstants.ENV_FLINK_OPT_DIR + " is set.",
+                System.getProperty(ConfigConstants.ENV_FLINK_OPT_DIR) != null);
```
Maybe we need to use `System.getenv`. Refer to `PackagedProgramUtils#getPythonJar`.

XComp replied:
Yeah, I thought of building a classpath that includes a mocked `flink-python.jar`. But adapting the enviroment variables is not advisable. That's were I stopped investigating.

wangyang0918 commented on flink-clients/src/test/java/org/apache/flink/client/program/PackagedProgramRetrieverImplTest.java:
```diff
@@ -215,20 +215,6 @@ public void testCreateWithUserLibDir() throws FlinkException {
     }
 
-    @Test
-    public void testCreateWithoutUserLibDir() throws IOException, FlinkException {
```
I could not fully understand why we need to remove the `testCreateWithoutUserLibDir` here. IIUC, it is used to verify the system classpath pickup.

XComp replied:
Not setting the user classpath makes the code fallback to the system classpath. We're not able to modify the system classpath after JVM is started already. Modifying `java.class.path` doesn't help here. The default system classpath contains multiple classes with main methods. That's why, the test would fail. The code path is, instead, covered by PackagedProgramRetrieverImplTest.testDeriveEntryClassInformationFromSystemClasspathExtractingTheJobClassFromThere`.

wangyang0918 commented on flink-clients/src/test/java/org/apache/flink/client/program/PackagedProgramRetrieverImplTest.java:
```diff
@@ -328,18 +329,15 @@ public void testFailIfJobDirDoesNotHaveEntryClass() {
+    // Right now, the test is failing because no error is thrown
+    @Ignore
+    @Test(expected = FlinkException.class)
+    public void testEntryClassNotFoundOnSystemClasspath() throws FlinkException {
```
IIUC, this test could be subsumed by `testDeriveEntryClassInformationFromSystemClasspathWithNonExistingJobClassName`.

XComp replied:
Correct, but I would keep it in as an additional test in case we decide to make the code more strict (as mentioned in [this comment](https://github.com/apache/flink/pull/16286#discussion_r658542629)).

Right now, I lean towards keeping it like it was implemented in the old implementation to not open any hidden problems which might make the PR become even bigger. Instead, I'd create a Flink issue to resolve that. But I wanted to get your opinions on that one. I wasn't 100% sure whether it's actually a desired behavior not checking the class existence on the system classpath.

wangyang0918 reviewed (commented):
@XComp Thanks for preparing this PR. It looks really good to me.

I just left some minor comments and please have a look.

zentol commented on flink-clients/src/main/java/org/apache/flink/client/program/PackagedProgramRetrieverImpl.java:
```diff
@@ -0,0 +1,214 @@
+
+        try {
+            return FromClasspathEntryClassInformationProvider.createFromSystemClasspath();
+        } catch (Throwable t) {
```
Why do you catch Throwables? Why is catching FlinkExceptions not sufficient?

XComp replied:
You're right. Initialliy, we were only expecting `IOException`, `NoSuchElementException`, and `IllegalArgumentException`. I, initially, thought that it would be good to have all errors covered. But I make it more explicit now. Additionally, I transform the `NoSuchElementException` and `IllegalArgumentException` into a checked `FlinkException` within the `FromClasspathEntryClassInformationProvider` to make these two error cases more prominent in the interface.

zentol replied:
You could've just catched `Exception` to cover these cases.

zentol commented on flink-clients/src/test/java/org/apache/flink/client/testjar/ClasspathProvider.java:
```diff
@@ -0,0 +1,230 @@
+
+        userDirHasTwoEntryClasses =
+                temporaryFolder.newFolder("_test_user_dir_has_two_entry_classes");
+        copyJar(JOB_JAR_PATH, userDirHasTwoEntryClasses);
```
Why do we need these copy operations? Couldn't we construct the classpath from the original locations?

XComp replied:
`PackagedProgramRetrieverImpl.createEntryClassInformationProvider` requires a File instance referring to the folder containing the files. That's why I kept the logic for creating the folders as it already was in the original code base.

zentol commented on flink-clients/src/test/java/org/apache/flink/client/program/PackagedProgramRetrieverImplTest.java:
```diff
@@ -0,0 +1,573 @@
+    @Test
+    public void testDeriveEntryClassInformationForCustomJar()
+            throws FlinkException, MalformedURLException {
+        // make loading from system classpath fail to make sure that it's not triggered
+        classpathProvider.setSystemClasspathWithTwoEntryClasses();
```
I don't understand why this would make it fail, nor what "it" that is not triggered actually is. Wouldn't it be more obvious if you did something like clearing the classpath?

XComp replied:
Good idea, that's more explicit. I agree.

zentol commented on flink-clients/src/test/java/org/apache/flink/client/program/PackagedProgramRetrieverImplTest.java:
```diff
@@ -0,0 +1,573 @@
+    }
+
+    @Test
+    public void testDeriveEntryClassInformationForCustomJar()
```
What is actually being derived from anything? It seems like all we're testing here is that the stuff passed via the method is eventually assigned.

XComp replied:
We're testing the code path where a jar is set. Admittedly, it's a quite basic test. But I tried to cover all code paths of `createEntryClassInformationProvider`.

zentol commented on flink-clients/src/test/java/org/apache/flink/client/program/PackagedProgramRetrieverImplTest.java:
```diff
@@ -0,0 +1,573 @@
+            throws IOException, FlinkException {
+        final EntryClassInformationProvider informationProvider =
+                PackagedProgramRetrieverImpl.createEntryClassInformationProvider(
+                        // the user directory must be specified
+                        classpathProvider.getUserDirectoryWithTwoEntryClasses(),
+                        // the user classpath is derived from the user directory outside of the
+                        // method
+                        classpathProvider.getURLUserClasspathWithTwoEntryClasses(),
```
Aren't these irrelevant since we explicitly provide the job classname anyway?

zentol replied:
the comments are also weird; they claim the user directory is required, but testDeriveEntryClassInformationFromSystemClasspathWithNonExistingJobClassName uses null for it and does the same assertions.

XComp replied:
There's a slight difference between providing a user classpath through the user directory and not providing one. In the former case, the code checks whether the job class appears in the user classpath. For the latter one, this check is not done.

That's the difference between `testDeriveEntryClassInformationFromClasspathWithJobClass` and `testDeriveEntryClassInformationFromSystemClasspathWithNonExistingJobClassName`.

zentol commented on flink-clients/src/test/java/org/apache/flink/client/program/PackagedProgramRetrieverImplTest.java:
```diff
@@ -0,0 +1,573 @@
+                PackagedProgramRetrieverImpl.createEntryClassInformationProvider(
+                        // the user directory must be specified
+                        classpathProvider.getUserDirectoryWithOneEntryClass(),
+                        // the user classpath is derived from the user directory outside of the
+                        // method
+                        classpathProvider.getURLUserClasspathWithEntryClass(),
```
which method are you referencing?

XComp replied:
The method that's being tested. I updated the comment to make this more clear

zentol commented on flink-clients/src/test/java/org/apache/flink/client/program/PackagedProgramRetrieverImplTest.java:
```diff
@@ -0,0 +1,573 @@
+    }
+
+    @Test
+    public void testTooManyEntryClassesOnUserClasspath() throws FlinkException, IOException {
```
So if 2 entry points are on the user CP we don't fail?

XComp replied:
No, it also fails if there are too many entrypoints on the classpath but no job class name is specified. But you're right. The test method's name is not specific enough. I updated the method name and additional tests. I realized that I wasn't precise here.

zentol commented on flink-clients/src/test/java/org/apache/flink/client/program/PackagedProgramRetrieverImplTest.java:
```diff
@@ -0,0 +1,573 @@
+        final Path workingDirectory = FileUtils.getCurrentWorkingDirectory();
+        final List<String> expectedClasspath = new ArrayList<>();
+        for (File file : classpathProvider.getUserDirectoryWithOneEntryClass().listFiles()) {
+            if (!file.getName().endsWith("jar")) {
+                // only jars are expected
+                continue;
+            }
```
Why should we silently ignore this? Wouldn't it be a bug in the ClassPathProvider?

XComp replied:
The `ClasspathProvider` provides all the resources that are on the classpath (including a text file in most of the cases). The production code filters them out in [PackagedProgramRetrieverImpl.extractUserClasspath](https://github.com/XComp/flink/blob/e9a2e57adf4f45958e6fa0e9e9f0f7afac50d01e/flink-clients/src/main/java/org/apache/flink/client/program/PackagedProgramRetrieverImpl.java#L208). The filtering here is just the corresponding test code checking this filtering.

XComp replied:
But I updated the condition to use `FileUtils` and added a more meaningful comment

zentol commented on flink-clients/src/test/java/org/apache/flink/client/program/PackagedProgramRetrieverImplTest.java:
```diff
@@ -0,0 +1,573 @@
+    }
+
+    @Test
+    public void testChildFirstDefaultConfiguration() throws FlinkException {
```
both this test and the one below should be renamed to clarify that they are about the class loaders

XComp replied:
Their actual purpose is to check whether the configuration is acknowledged. The classloader is only the vehicle to test that. That's why, I would object renaming them. Or do you have some proposal for a better name? 🤔

zentol replied:
Shouldn't this then belong into a `PackagedProgramTest`?

XComp replied:
The actual bug which we want to fix in `FLINK-21445` is that we missed setting the configuration in the `ClassPathPackagedProgramRetriever` which is now replaced by `PackagedProgramRetrieverImpl`. That's why I added the tests here: To cover this behavior

zentol commented on flink-clients/src/test/java/org/apache/flink/client/deployment/application/FromClasspathEntryClassInformationProviderTest.java:
```diff
@@ -0,0 +1,173 @@
+    }
+
+    @Test
+    public void testJarFromSystemClasspathSanityCheck() {
```
what do we need this test for? If you want to check that we can properly load something from the system CP without any modifications to it, then why not put a jar on the test classpath and try loading a job from that.

XComp replied:
That's a sanity check I copied over from the old `ClassPathPackagedProgramRetrieverTest`. I thought it's reasonable to check whether the jar loading without modifying the `java.class.path` System property. Isn't that essentially the same as what you suggest? ...just with less code?

zentol replied:
We can keep it because it already exists, but what I objected to is searching for the junit jar on a conceptual level. It is not sound to couple the usage of some testing library to the correctness of production code.

XComp replied:
You have a point. I agree. But I would also vote for keeping it like this for now as I want to avoid opening another refactoring sub-project.

XComp commented on flink-clients/src/test/java/org/apache/flink/client/deployment/application/JarManifestParserTest.java:
```diff
@@ -157,6 +158,24 @@ public void testFindOnlyEntryClassMultipleJarsWithSingleManifestEntry() throws I
+        Optional<String> optionalValue =
+                JarManifestParser.findFirstManifestAttribute(
+                        TestJob.getTestJobJar(),
+                        "some-attr",
```
I added the tests because I made the `JarManifestParser.findFirstManifestAttribute` publicly accessible. Hence, I thought of adding test cases for it might be good as well.

zentol replied:
but that is this specific attribute for?

XComp replied:
Fair enough - it doesn't add any value but might confuse the reader.

XComp commented on flink-yarn/src/main/java/org/apache/flink/yarn/entrypoint/YarnApplicationClusterEntryPoint.java:
```diff
@@ -129,16 +129,12 @@ private static PackagedProgramRetriever getPackagedProgramRetriever(
-                        .setJarFile(userApplicationJar)
-                        .setJobClassName(jobClassName);
-        return retrieverBuilder.build();
+        return PackagedProgramRetrieverImpl.create(
+                userLibDir, userApplicationJar, jobClassName, programArguments, configuration);
```
I switched to actual factory methods because we didn't use the builder functionality. `@Nullable` values are passed.  The production code boiled down to two different signatures reflected in the `PackagedProgramRetrieverImpl.create` factory methods.

XComp commented on flink-clients/src/main/java/org/apache/flink/client/program/PackagedProgramRetrieverImpl.java:
```diff
@@ -0,0 +1,214 @@
+     */
+    public static PackagedProgramRetrieverImpl create(
+            @Nullable File userLibDir,
+            @Nullable File jarFile,
```
I'm not 100% sure whether I understand correctly. You're referring to `jarFile`? My goal was to have the `createEntryClassInformationProvider` method contain all the different code paths to select the right `EntryClassInformationProvider`. That enabled us to test this method specifically and be as close as possible to the original code. Making one of the fields `@Nonnull` would have meant to move some if condition out of this method.

XComp commented on flink-clients/src/test/java/org/apache/flink/client/deployment/application/FromJarEntryClassInformationProviderTest.java:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    @Test
+    public void testPythonJarFile() {
```
I guess, you're right. It was just that I thought it's an obvious test and just later on realized that it's not that easy to fix. Reconsidering this, I have to admit that my approach is not the right one. I'm gonna revert the tests and comment on `FLINK-23154` accordingly.

XComp commented:
Thanks @SteNicholas , @wangyang0918 and @zentol for your detailed feedback. I addressed all the comments or commented on specific items. Please have another look.

XComp commented on flink-clients/src/test/java/org/apache/flink/client/program/PackagedProgramRetrieverImplTest.java:
```diff
@@ -590,4 +567,25 @@ private JobGraph retrieveJobGraph(
+            throws MalformedURLException {
+        Preconditions.checkArgument(
+                directory.listFiles() != null,
+                "The passed File does not seem to be a directory: " + directory.getAbsolutePath());
```
I initially thought the same. But objected it because in the end we're using `listFiles` in the code. And `listFiles` does not only return `null` in case of it not being a directory. But I udpated the error message to reflect that.

XComp commented on flink-clients/src/test/java/org/apache/flink/client/program/PackagedProgramRetrieverImplTest.java:
```diff
@@ -126,10 +126,9 @@ public void testDeriveEntryClassInformationForPythonBasedOnParameter() throws Fl
-        multipleEntryClassesClasspathProvider.setSystemClasspath();
+    public void testDeriveEntryClassInformationForCustomJar() throws FlinkException {
+        // clearing the system classpath to make sure that no data is collected from there
+        System.setProperty("java.class.path", "");
```
Good point. I will update it...

XComp commented on flink-clients/src/test/java/org/apache/flink/client/testjar/ClasspathProvider.java:
```diff
@@ -162,6 +164,7 @@ public void before() throws IOException {
 
         directory = temporaryFolder.newFolder(directoryNameSuffix);
         directoryContentCreator.accept(directory);
+        resetSystemClasspath();
```
You're right. I just felt that it would be another safety net. But ok, I'm gonna remove it again since I have no actual reason to keep it.

XComp commented on flink-clients/src/test/java/org/apache/flink/client/program/PackagedProgramRetrieverImplTest.java:
```diff
@@ -197,8 +197,8 @@ public void testDeriveEntryClassInformationFromClasspathWithJobClass()
                         multipleEntryClassesClasspathProvider.getDirectory(),
-                        // the user classpath is derived from the user directory outside of the
-                        // method
+                        // the user classpath is derived from the user directory outside of
+                        // createEntryClassInformationProvider
```
I find it a bit confusing to pass in the user directory and the user classpath. But this was required to match the original functionality. That's why I added the comment to make it obvious that the user classpath and the user directory are actually referring to the same thing.

zentol commented on flink-clients/src/main/java/org/apache/flink/client/program/PackagedProgramRetrieverImpl.java:
```diff
@@ -0,0 +1,213 @@
+            return FromJarEntryClassInformationProvider.createFromCustomJar(jarFile, jobClassName);
+        }
+
+        if (userLibDir != null) {
```
this is a bit weird isn't it. Couldn't we make `userClasspath` nullable instead? And move the null check for the userlib to the `create` variant that accepts it? That would also address https://github.com/apache/flink/pull/16286/files/a7a79b1944b91a03c7567ce29c561186b1f13d74..c071a23a319c89310e19a40f6d85217936ab1067#r662129429.

XComp replied:
I don't know why I didn't see the obvious. 🤦 But yeah, that's way better.

XComp commented on flink-clients/src/main/java/org/apache/flink/client/program/PackagedProgramRetrieverImpl.java:
```diff
@@ -0,0 +1,215 @@
+
+        return fromSystemClasspath(jobClassName);
+    }
+
+    static EntryClassInformationProvider fromSystemClasspath(@Nullable String jobClassName)
```
Good catch. That one I must have missed. But the `@VisibleForTesting` is actually obsolete and the visibility of `fromSystemClasspath` and `fromUserClasspath` wrong. The testing is entirely done on the calling method `createEntryClassInformationProvider`. I made `fromSystemClasspath` and `fromUserClasspath` private and removed the `@VisibleForTesting` annotation

XComp commented on flink-clients/src/main/java/org/apache/flink/client/deployment/application/FromClasspathEntryClassInformationProvider.java:
```diff
@@ -0,0 +1,211 @@
+     * @param classpath The classpath the job class is expected to be part of.
+     * @return The {@code FromClasspathEntryClassInformationProvider} providing the job class found
+     *     on the passed classpath.
+     * @throws IOException If some Jar listed on the classpath wasn't accessible.
```
Good catch. I moved the IOException handling out of the private class into `PackagedProgramRetrieverImpl` analogous to how it's done for the other classpath accesses.

XComp commented on flink-clients/src/main/java/org/apache/flink/client/deployment/application/EntryClassInformationProvider.java:
```diff
@@ -0,0 +1,42 @@
+    Optional<File> getJarFile();
+
+    /**
+     * Returns the name of the job class.
```
I updated the JavaDoc in the implementing classes. That way the documentation is closer to the actual implementation.

XComp commented:
I rebased the PR to make the `JobMasterITCase` succeed in the AzureCI builds

zentol commented on flink-clients/src/main/java/org/apache/flink/client/deployment/application/FromJarEntryClassInformationProvider.java:
```diff
@@ -68,11 +68,33 @@ private FromJarEntryClassInformationProvider(
+    /**
+     * Returns the specified job class name that is either available in the corresponding {@code
+     * jarFile} or, if the {@code jarFile} is not set, should be assumed to be contained in the
+     * system classpath. It can return an emtpy {@code Optional} if the respective {@code jarFile}
```
emtpy -> empty

zentol commented on flink-clients/src/main/java/org/apache/flink/client/deployment/application/FromJarEntryClassInformationProvider.java:
```diff
@@ -68,11 +68,33 @@ private FromJarEntryClassInformationProvider(
 
+    /**
+     * Returns the specified {@code jarFile}. If no {@code jarFile} was specified it can only return
+     * an empty {@code Optional} if the {@code jobClassName} is specified instead. In that case, the
```
The condition when it returns an empty optional is unnecessary and just creates confusion. The jar _and_ classname not being specified is a case that a consumer of this APIn ever has to be concerned with because the constructor rejects it.

XComp replied:
Fair enough, the information is also given in `@return` docs. So, I guess, that's good enough.

zentol commented on flink-clients/src/main/java/org/apache/flink/client/deployment/application/FromJarEntryClassInformationProvider.java:
```diff
@@ -0,0 +1,102 @@
+import java.util.Optional;
+
+/**
+ * {@code FromJarEntryClassInformationProvider} is used for cases where the Jar archive is
+ * explicitly specified.
```
The jar doesn't have to be explicitly specified.

XComp replied:
I thought about it once more. I got the precondition from [PackagedProgram:687](https://github.com/apache/flink/blob/c08bef1f7913eb1c416c278354fd62b82b172549/flink-clients/src/main/java/org/apache/flink/client/program/PackagedProgram.java#L687) not realizing that the `FromJarEntryClassInformationProvider` is actually on used for the PyFlink (where the Jar and the job class name are explicitly given) or [if a jarFile is specified in DefaultPackagedProgramRetriever](https://github.com/XComp/flink/blob/FLINK-21445/flink-clients/src/main/java/org/apache/flink/client/program/DefaultPackagedProgramRetriever.java#L128). Hence, we can make the implementation more stricter which makes the JavaDoc valid again.

zentol commented on flink-clients/src/main/java/org/apache/flink/client/deployment/application/FromJarEntryClassInformationProvider.java:
```diff
@@ -0,0 +1,102 @@
+    /**
+     * Returns the specified job class name that is either available in the corresponding {@code
+     * jarFile} or, if the {@code jarFile} is not set, should be assumed to be contained in the
+     * system classpath. It can return an emtpy {@code Optional} if the respective {@code jarFile}
+     * is set. In that case one should assume the job class being the entry class of the respective
```
```suggestion
     * system classpath. It can return an empty {@code Optional} despite the respective {@code jarFile}
     * being set if the job class is the entry class of the jar.
```
The descriptions are bit difficult to understand :/

Note also that this behavior is super weird and we should just return the entry class imo.

XComp replied:
I agree, but that would not reflect the original code. That's why I didn't add it. We also have to consider the fact that the entry class is not specified in the jar archive. In that case setting the job class name explicitly makes sense

zentol commented on flink-clients/src/main/java/org/apache/flink/client/deployment/application/FromJarEntryClassInformationProvider.java:
```diff
@@ -0,0 +1,102 @@
+     * {@code jarFile}.
+     *
+     * @return Returns the job class that can either be found in the respective {@code jarFile} or
+     *     on the system classpath. It can also return an empty {@code Optional} if the {@code
+     *     jarFile} is specified. In that case the caller should assume the job class being provided
+     *     by the {@code jarFile} as the entry class.
```
similar suggestion to above

zentol commented on flink-clients/src/main/java/org/apache/flink/client/program/PackagedProgramRetrieverImpl.java:
```diff
@@ -0,0 +1,215 @@
+import java.util.stream.Collectors;
+
+/** {@code PackageProgramRetrieverImpl} implements {@link PackagedProgramRetriever}. */
+public class PackagedProgramRetrieverImpl implements PackagedProgramRetriever {
```
Recently we used `Default*` for the default interface implementations.

zentol commented on flink-clients/src/main/java/org/apache/flink/client/program/PackagedProgramRetrieverImpl.java:
```diff
@@ -0,0 +1,215 @@
+import java.util.NoSuchElementException;
+import java.util.stream.Collectors;
+
+/** {@code PackageProgramRetrieverImpl} implements {@link PackagedProgramRetriever}. */
```
Well this is quite a non-comment isn't it. It could say something like "default X implementation that can retrieve a packaged program from a specific jar or the system classpath."

zentol commented on flink-clients/src/main/java/org/apache/flink/client/program/PackagedProgramRetrieverImpl.java:
```diff
@@ -0,0 +1,215 @@
+     *
+     * @param userLibDir The user library directory that is used for generating the user classpath
+     *     if specified. The system classpath is used if not specified.
+     * @param jarFile The jar archive used expecting to have the job class included.
```
```suggestion
     * @param jarFile The jar archive expected to contain the job class; null if the job class is on the system classpath.
```

zentol commented on flink-clients/src/main/java/org/apache/flink/client/program/PackagedProgramRetrieverImpl.java:
```diff
@@ -0,0 +1,215 @@
+     * @param userLibDir The user library directory that is used for generating the user classpath
+     *     if specified. The system classpath is used if not specified.
+     * @param jarFile The jar archive used expecting to have the job class included.
+     * @param jobClassName The job class that will be used if specified. The classpath is used to
+     *     detect any main class if not specified.
```
```suggestion
     * @param jobClassName The job class to use; if null the system classpath will be scanned for possible main classes.

XComp commented:
Thanks @zentol . I addressed your JavaDoc comments and took the opportunity to make the condition on the `FromJarEntryClassInformationProvider` stricter. It's only called with a jar archive being specified. That resolved also some of your JavaDoc-related comments.

wangyang0918 reviewed (approved):
@XComp Thanks for addressing the comments. This PR looks good to me.
+1 for merging.

XComp commented on flink-clients/src/main/java/org/apache/flink/client/deployment/application/FromJarEntryClassInformationProvider.java:
```diff
@@ -0,0 +1,92 @@
+     */
+    public static FromJarEntryClassInformationProvider createFromCustomJar(
+            File jarFile, @Nullable String jobClassName) {
+        return new FromJarEntryClassInformationProvider(jarFile, jobClassName);
```
I rather don't want to change the failure logic as part of this PR. Just by looking at the code, it would be possible, I guess. But I'm hesitant to do this because it might open other refactoring efforts which I want to avoid.

wangyang0918 replied:
Make sense.

XComp commented:
I reorganized the commits squashing most of them together and rebased the branch to the most recent `master`

XComp commented:
Backport PRs for `release-1.11` (here I only applied the configuration fix excluding the refactoring), `release-1.12` and `release-1.13` are created.
